### DISCUSSION TITLE
C++: Add suport for asynchronous direct IO

### DISCUSSION THREAD

**Discussion by olesalscheider** [foxglove/mcap#1087] [General] [answered]
We are working with large messages at high frequencies which result in high data throughput. When trying to store this data using the mcap format, we hit unexpected limitations with some of our storage configurations. Especially with some larger enterprise NVME SSDs like Samsung's PM9A3 (15.36TB version) we noticed frame drops during recording even though we were far from the theoretical maximum throughput.

I did some benchmarks and found that normal buffered synchronous writes only achieve around 1530 MB/s which is much less than the 4000 MB/s mentioned in the datasheet. With direct asynchronous writes (queue depth of 2048) I could achieve 3940 MB/s which is much closer to the advertised performance.

It would be great if mcap could support direct asynchronous writes so that it becomes a suitable storage backend in these high-bandwidth scenarios.

**Comment by james-rms** [comment]
It depends a little on what waiting behaviour you'd want. You could easily write a  your own`mcap::IWritable` implementation which executes asynchronous writes and never waits on them to complete, which would probably maximize throughput. However, you wouldn't have any guarantees about when data makes it to the disk, until you close the file.

**Reply by olesalscheider** [reply]
I think not having guarantees is acceptable in many use-cases, and it is with ours. And you can still get some guarantees / control by limiting the number of write requests in flight, the total amount of outstanding bytes to be written or the maximum age of the outstanding write requests - if you specify limits and one is violated the write function could block or return an error code.

For this to work the interface of `mcap::IWritable` however would have to be extended by a non-blocking write call. It should probably have a callback argument or some other way to notify the user when the write operation is completed. This is necessary so that the user knows when the memory can be freed or returned to the middleware (in case of loaned messages for zero copy).

**Reply by james-rms** [reply, marked as answer]
Seems like a reasonable desire; I'm not convinced you'd need to add another virtual method to `IWritable` though. The MCAP library doesn't need to provide a callback to the IWritable to do its job. 

You could pass your callback to your subclass some other way (either in the constructor or via another non-virtual method that you call before every message write), and have the standard `write()` call behave however you need it to.

**Reply by olesalscheider** [reply]
Hm, yes... What I don't like about the constructor is that I don't know how we can tell the user which write command was completed. Do we identify them just by the pointer to the data buffer? But who knows, maybe someone has some use case where he wants the same buffer to be written twice in certain situations. So providing some user data with the callback / having different (lambda function) callbacks sounds useful.

For each write calling a function to set the callback / user data and then calling the write function would be possible, of course. I'll look into that.
I'm not sure if this is something that other people would be interested in for mcap / rosbag2. Having a nice API around it in McapWriter might be nice then.
